src/main/java/org/opensearch/neuralsearch/search/collector/HybridTopFieldDocSortCollector.java

```diff
@@ -0,0 +1,407 @@
+                fieldComparator.setTopValue(after.fields[k]);
+            }
+        }
+    }
+
+    /*
+     TopFieldDocs per subquery
+    */
+    private TopFieldDocs topDocsPerQuery(
+        int start,
+        int howMany,
+        PriorityQueue<FieldValueHitQueue.Entry> pq,
+        int totalHits,
+        SortField[] sortFields
+    ) {
+        if (howMany < 0) {
+            throw new IllegalArgumentException(
+                String.format(Locale.ROOT, "Number of hits requested must be greater than 0 but value was %d", howMany)
+            );
+        }
+
+        if (start < 0) {
+            throw new IllegalArgumentException(
+                String.format(Locale.ROOT, "Expected value of starting position is between 0 and %d, got %d", howMany, start)
+            );
+        }
+
+        if (start >= howMany || howMany == 0) {
+            return EMPTY_TOPDOCS;
+        }
+
+        int size = howMany - start;
+        ScoreDoc[] results = new ScoreDoc[size];
+
+        // Get the requested results from pq.
+        populateResults(results, size, pq);
+
+        return new TopFieldDocs(new TotalHits(totalHits, totalHitsRelation), results, sortFields);
+    }
+
+    /*
+      Results are converted in the FieldDocs and the value of the field on which the sorting is applied has been added in the FieldDoc.
+     */
+    private void populateResults(ScoreDoc[] results, int howMany, PriorityQueue<FieldValueHitQueue.Entry> pq) {
+        FieldValueHitQueue<FieldValueHitQueue.Entry> queue = (FieldValueHitQueue<FieldValueHitQueue.Entry>) pq;
+        for (int i = howMany - 1; i >= 0 && pq.size() > 0; i--) {
+            // adding to array if index is within [0..array_length - 1]
+            if (i < results.length) {
+                FieldValueHitQueue.Entry entry = queue.pop();
+                final int n = queue.getComparators().length;
+                final Object[] fields = new Object[n];
+                for (int j = 0; j < n; ++j) {
+                    fields[j] = queue.getComparators()[j].value(entry.slot);
+                }
+
+                results[i] = new FieldDoc(entry.doc, entry.score, fields);
+            }
+        }
+    }
+
+    // Add the entry in the Priority queue
+    private void add(int slot, int doc, FieldValueHitQueue<FieldValueHitQueue.Entry> compoundScore, int subQueryNumber, float score) {
+        FieldValueHitQueue.Entry bottomEntry = new FieldValueHitQueue.Entry(slot, docBase + doc);
+        bottomEntry.score = score;
+        bottom = compoundScore.add(bottomEntry);
+        // The queue is full either when totalHits == numHits (in SimpleFieldCollector), in which case
+        // slot = totalHits - 1, or when hitsCollected == numHits (in PagingFieldCollector this is hits
+        // on the current page) and slot = hitsCollected - 1.
+        assert slot < numHits;
+        queueFull[subQueryNumber] = slot == numHits - 1;
```

Comment:
nit: I would rather put this in a more expanded form for better readability

Reply:
ack